modules/bls-dkg/index.js

```diff
@@ -43,6 +46,96 @@ function bigIntToBuffer(bigint) {
   return hexToBuffer(bigint.toString(16))
 }
 
+function bigIntToBufferBE(bigint, bytes) {
+  let v = bigint.toString(16);
+  v = '0'.slice(0, v.length % 2) + v;
+  const buf = Buffer.from(v, 'hex');
+  if (bytes && buf.length < bytes) {
+    return Buffer.concat([Buffer.alloc(bytes - buf.length), buf]);
+  }
+  return buf;
+}
+
+/** Integer to Octet Stream */
+function i2osp(value, length) {
+  if (value < 0 || value >= 1n << BigInt(8 * length)) {
+    throw new Error(`bad I2OSP call: value=${value} length=${length}`);
+  }
+  const res = Array.from({ length }).fill(0);
+  for (let i = length - 1; i >= 0; i--) {
+    res[i] = value & 0xff;
+    value >>>= 8;
+  }
+  return new Uint8Array(res);
+}
+
+/** Concatenate multiple buffers into one */
+function concatBytes(...arrays) {
+  if (arrays.length === 1) {
+    return arrays[0];
+  }
+  const length = arrays.reduce((a, arr) => a + arr.length, 0);
+  const result = new Uint8Array(length);
+  for (let i = 0, pad = 0; i < arrays.length; i++) {
+    const arr = arrays[i];
+    result.set(arr, pad);
+    pad += arr.length;
+  }
+  return result;
+}
+
+/** Get the indices from a sub key path string */
+function getSubKeyPathIndices(path) {
+  return path.replace(/^m\//, '')
+    .split('/')
+    .map((index) => parseInt(index, 10));
+}
+
+/** Calculate offset from a public key */
+function calculateOffset(pk, index, chaincode) {
+  const salt = i2osp(index, 4);
+  const buffer = concatBytes(pk, salt);
+  return createHmac('sha512', bigIntToBufferBE(chaincode, 32)).update(buffer).digest();
+}
+
+/** Derive a public key */
+function publicDerive(pk, path, chaincode) {
```

Comment:
More of a nit, but the signature of `publicDerive` and `privateDerive` are a bit unconventional. Would be nicer if the key and the chain parameters were together (ie a "keychain"): `function publicDerive(pk, chaincode, path)`

Reply:
You are right, thanks!